### PR TITLE
[AnimComponent] Public API and engine examples

### LINKED ISSUE #2533: Create new animation engine examples
The anim component needs engine examples before it can be made public.

The following functionality should be covered:
- Locomotion example
- Anim clips examples for animating entity component properties
- Blend Trees - 1D
- Blend Trees - 2D Directional
- Blend Trees - 2D Cartesian
- Blend Trees - Direct Blending

PCUI should be used to handle any interaction with the examples.

### PR DESCRIPTION
**Pull request by ellthompson**
Makes all the necessary API to use the AnimComponent and its related classes public.

Creates the following engine examples to showcase the new anim component (Fixes #2533 ):
- Locomotion
- Animating entity component properties
- 1D blending
- 2D blending (cartesian)
- 2D blending (directional)

Also updates the anim binder to use the current entity as the root of the animation hierarchy for both model and render components. This root can be changed by the user using the AnimComponent#rootBone property.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: examples/animation/blend-trees-1d.html
Comment (Maksims) [inline examples/animation/blend-trees-1d.html:126]: This syntax asks for a better solution. Perhaps small jobs counter with a callback is a better solution to handle this callback pyramid.

File: src/framework/components/anim/controller.js
Comment (Maksims) [inline src/framework/components/anim/controller.js:185]: This is a very slow way to deeply compare data and leads to allocations. If the structure of the provided values known, then it is better to write a specific comparison method, which fails early if deviations are found.

**Comment by Maksims** [comment]
Where do assets (model and animations) come from? Do we have a license/right to post them without an attached license and without infringing copyrights?

File: src/framework/components/anim/controller.js
Comment (ellthompson) [reply src/framework/components/anim/controller.js]: Good catch here. I've addressed this in another PR which will go in before this one.

**Comment by ellthompson** [comment]
> Where do assets (model and animations) come from? Do we have a license/right to post them without an attached license and without infringing copyrights?
They're from mixamo. We investigated this and it does look like mixamo assets would pose legal issues. I've switched these assets out now.

**Comment by willeastcott** [comment]
Can you rename `Idle.glb` to `idle.glb` please? (Wow, I'm super-picky!) 😆

**Comment by mvaligursky** [comment]
Considering we're about to release engine side of the render component, perhaps we should change these examples to use it?

**Comment by willeastcott** [comment]
Yeah, I'd agree with that.

File: examples/animation/blend-trees-1d.html
Comment (mvaligursky) [inline examples/animation/blend-trees-1d.html:76]: what does this do? It's not something we seem to use in other engine examples

File: examples/animation/blend-trees-1d.html
Comment (mvaligursky) [inline examples/animation/blend-trees-1d.html:127]: perhaps use render component

File: examples/animation/blend-trees-2d-cartesian.html
Comment (mvaligursky) [inline examples/animation/blend-trees-2d-cartesian.html:145]: use something like this instead

```            
var entity = model.resource.instantiateRenderEntity({
        castShadows: true
});
```

File: examples/animation/blend-trees-directional.html
Comment (mvaligursky) [inline examples/animation/blend-trees-directional.html:153]: use something like this

```            
var entity = asset.resource.instantiateRenderEntity({
        castShadows: true
});
```

File: examples/animation/component-properties.html
Comment (mvaligursky) [inline examples/animation/component-properties.html:55]: just change "model" to "render"

File: examples/animation/component-properties.html
Comment (mvaligursky) [inline examples/animation/component-properties.html:74]: here as well, use render

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:245]: use render component

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:262]: "render"

Comment (mvaligursky) [reply examples/animation/locomotion.html:262]: actually, this might not work if it's collider.

File: examples/examples.js
Comment (mvaligursky) [inline examples/examples.js:8]: sort alphabetically

**Comment by kungfooman** [comment]
In `src/framework/entity.js` the `AnimComponent` needs to be declared as property, like `AnimationComponent`, otherwise it won't be in `playcanvas.d.ts`:

```js
* @property {AnimationComponent} [animation] Gets the {@link AnimationComponent} attached to this entity. [read only]
* @property {AnimComponent} [anim] Gets the {@link AnimComponent} attached to this entity. [read only]
```

**Comment by kungfooman** [comment]
In `src/framework/components/anim/component.js` you only declare two properties:

```js
 * @property {number} speed Speed multiplier for animation play back speed. 1.0 is playback at normal speed, 0.0 pauses the animation.
 * @property {boolean} activate If true the first animation will begin playing when the scene is loaded.
```

But the component has four properties:

https://github.com/playcanvas/engine/blob/cbacd4c4bcca5b87e650825b57309f4aaa62e691/src/framework/components/anim/system.js#L38-L47

(lacking `enabled` and `playing` in JSDoc)

File: src/framework/components/anim/system.js
Comment (kungfooman) [inline src/framework/components/anim/system.js:51]: Not public without JSDoc, same for `initializeComponentData` method right above

**Comment by ellthompson** [comment]
> In `src/framework/entity.js` the `AnimComponent` needs to be declared as property, like `AnimationComponent`, otherwise it won't be in `playcanvas.d.ts`:
> 
> ```js
> * @property {AnimationComponent} [animation] Gets the {@link AnimationComponent} attached to this entity. [read only]
> * @property {AnimComponent} [anim] Gets the {@link AnimComponent} attached to this entity. [read only]
> ```

Thanks, i'll add this in.

File: src/framework/components/anim/system.js
Comment (ellthompson) [reply src/framework/components/anim/system.js:51]: These methods aren't public on any other components, so i'll keep them private for now.

**Comment by ellthompson** [comment]
> In `src/framework/components/anim/component.js` you only declare two properties:
> 
> ```js
>  * @property {number} speed Speed multiplier for animation play back speed. 1.0 is playback at normal speed, 0.0 pauses the animation.
>  * @property {boolean} activate If true the first animation will begin playing when the scene is loaded.
> ```
> 
> But the component has four properties:
> 
> https://github.com/playcanvas/engine/blob/cbacd4c4bcca5b87e650825b57309f4aaa62e691/src/framework/components/anim/system.js#L38-L47
> 
> (lacking `enabled` and `playing` in JSDoc)

I'll expose playing here but enabled doesn't seem to be listed in the JSDocs of other components.

File: src/framework/components/anim/system.js
Comment (kungfooman) [reply src/framework/components/anim/system.js:51]: Is PC Viewer bad style then? It is used publicly here: https://github.com/playcanvas/playcanvas-viewer/blob/master/src/viewer.ts#L802

All these private/non-documented methods cause dozens of type errors across projects.

Comment (ellthompson) [reply src/framework/components/anim/system.js:51]: Yeah I should really have added a comment at that line in the viewer. It's not a pattern we publicly support but it was useful for the viewers rather particular timeline scrubber.

File: examples/animation/blend-trees-1d.html
Comment (ellthompson) [reply examples/animation/blend-trees-1d.html]: It does. There's plans in the not too distance future to rewrite these examples. I think that'd be a good time to rethink all of the asset loading for our examples.

File: examples/animation/blend-trees-1d.html
Comment (mvaligursky) [inline examples/animation/blend-trees-1d.html:89]: seems unused?

File: examples/animation/blend-trees-2d-cartesian.html
Comment (mvaligursky) [inline examples/animation/blend-trees-2d-cartesian.html:126]: perhaps update examples to use this when merged
https://github.com/playcanvas/engine/pull/2978

File: examples/animation/blend-trees-2d-cartesian.html
Comment (mvaligursky) [inline examples/animation/blend-trees-2d-cartesian.html:97]: is this used?

File: examples/animation/blend-trees-2d-cartesian.html
Comment (mvaligursky) [inline examples/animation/blend-trees-2d-cartesian.html:144]: it'd be good to add explanations, especially to anim related code, as it acts as tutorial for people

File: examples/animation/blend-trees-2d-cartesian.html
Comment (mvaligursky) [inline examples/animation/blend-trees-2d-cartesian.html:159]: comments here too please

File: examples/animation/blend-trees-directional.html
Comment (mvaligursky) [inline examples/animation/blend-trees-directional.html:105]: probably not used here as well

File: examples/animation/blend-trees-directional.html
Comment (mvaligursky) [inline examples/animation/blend-trees-directional.html:196]: I'm not sure arrow functions are ES5 / IE11?

File: examples/animation/blend-trees-directional.html
Comment (mvaligursky) [inline examples/animation/blend-trees-directional.html:212]: position.y *= -1;

File: examples/animation/component-properties.html
Comment (mvaligursky) [inline examples/animation/component-properties.html:137]: even some comments on the json format could be good

File: examples/animation/component-properties.html
Comment (mvaligursky) [inline examples/animation/component-properties.html:187]: comments please

File: examples/animation/component-properties.html
Comment (mvaligursky) [inline examples/animation/component-properties.html:338]: can this be passed to constructor above?

File: examples/animation/component-properties.html
Comment (mvaligursky) [inline examples/animation/component-properties.html:358]: missing new line at the end

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:146]: you should probably use pc.ANIM_LESS_THAN and similar in these instead of strings, as that's the API, similar to how you use pc.ANIM_PARAMETER_INTEGER few lines down.

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:183]: maybe unused?

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:235]: comments, but specifically why you created additional parent entity

File: examples/animation/locomotion.html
Comment (mvaligursky) [reply examples/animation/locomotion.html:258]: actually yes, it should be a render.
the collider is not mesh collider, so it does not use the model

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:343]: you could just use camera position for the near one

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:352]: comments please

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:368]: any specific reason this is postUpdate and not update? explain in the comment

File: examples/animation/locomotion.html
Comment (mvaligursky) [inline examples/animation/locomotion.html:409]: currentPosition.copy(targetPosition);

File: src/resources/anim-clip.js
Comment (mvaligursky) [inline src/resources/anim-clip.js:8]: maybe this should stay private? There are no public properties of functions, so why the class.

File: src/framework/components/anim/component.js
Comment (mvaligursky) [inline src/framework/components/anim/component.js:371]: maybe print a warning in debug mode that root bone cannot be found here if appropriate

Comment (ellthompson) [reply src/framework/components/anim/component.js:371]: Done 👍

File: examples/animation/locomotion.html
Comment (ellthompson) [reply examples/animation/locomotion.html:343]: It could, I just like the idea of the ray connecting two points on a frustum volume.

File: examples/animation/locomotion.html
Comment (ellthompson) [reply examples/animation/locomotion.html:353]: For some reason it's not showing up but i've added this in.

File: examples/animation/locomotion.html
Comment (mvaligursky) [reply examples/animation/locomotion.html:343]: in that case use front and back culling distances set on the camera instead of 0.1 and 1000

File: examples/animation/blend-trees-2d-cartesian.html
Comment (ellthompson) [reply examples/animation/blend-trees-2d-cartesian.html:124]: Will do 👍

File: examples/animation/locomotion.html
Comment (ellthompson) [reply examples/animation/locomotion.html:343]: Sounds good 👍

File: examples/animation/blend-trees-2d-cartesian.html
Comment (ellthompson) [reply examples/animation/blend-trees-2d-cartesian.html]: I've switched these examples over to use asset loader function. I've also updated the function to handle data asset types too. Could you take a look?

File: examples/animation/blend-trees-2d-cartesian.html
Comment (ellthompson) [reply examples/animation/blend-trees-2d-cartesian.html:171]: I've pointed these parts of the examples to the PCUI documentation.

File: examples/assets/scripts/asset-loader.js
Comment (mvaligursky) [inline examples/assets/scripts/asset-loader.js:24]: we have this code twice now .. can it be wrapped in some shared function.

File: src/framework/components/anim/component.js
Comment (mvaligursky) [inline src/framework/components/anim/component.js:371]: I'd probably change this to simply
`this.data.rootBone = entity`
so that it gets set to null if not found .. otherwise the previous value will stay assigned? and do 
`if (!entity) console.warning ...`

**Comment by mvaligursky** [comment]
Looks good to me now, two more comments to address, but approving. Perhaps wait for one more approval. (@willeastcott )

**Review by mvaligursky — APPROVED** [review]

File: src/framework/components/anim/component.js
Comment (mvaligursky) [inline src/framework/components/anim/component.js:372]: put the if inside the ifdef

**Review by willeastcott — APPROVED** [review]

**Pull request merged**
